hsm/hsm.go

```diff
@@ -0,0 +1,291 @@
+package hsm
+
+import (
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/google/uuid"
+	"github.com/miekg/pkcs11"
+	"github.com/spf13/viper"
+)
+
+var ErrNoHSM = errors.New("no HSM configured")
+var ErrInvalidParam = errors.New("invalid input parameter")
+var ErrInvalidFilename = errors.New("invalid pkcs#11 module path/filename")
+var ErrInitFailure = errors.New("pkcs#11 module init failure")
+var ErrMissingSlot = errors.New("no slot defined")
+var ErrSessionFailure = errors.New("unable to open session with slot 0")
+var ErrLoginFailure = errors.New("unable to login to slot 0 with the specific pin")
+var ErrQueryFailure = errors.New("unable to execute the requested PKCS#11 api call")
+
+var CA_KEY_LABEL = "LITE Root CA"
+var CA_KEY_UUID = []byte{0xd0, 0x61, 0x9e, 0x62, 0xdd, 0xa2, 0x43, 0xb4, 0xb5, 0x3c, 0x85, 0x0b, 0x07, 0xf0, 0x78, 0x1c}
+
+// An HSM manages a session to a single HSM instance.  Generally,
+// there will only be a single of these instances for a given
+// execution of the program.
+type HSM struct {
+	api     *pkcs11.Ctx
+	session pkcs11.SessionHandle
+}
+
+// NewHSM will attempt to connect to the configured HSM.  It will
+// return ErrNoHSM if no HSM has been configured. Other errors
+// indicate an actual error connecting to the HSM.  `Close()` should
+// be called before exiting, when done with the HSM
+func NewHSM() (*HSM, error) {
+	// Cleanup function, needed due to lack of errdefer in Go.
+	cleanup := func() {}
+	module := viper.GetString("hsm.module")
+	pin := viper.GetString("hsm.pin")
+
+	if module == "" || pin == "" {
+		return nil, ErrNoHSM
+	}
+
+	// Make sure the specified PKCS#11 shared module actually
+	// exists
+	_, err := os.Stat(module)
+	if err != nil {
+		return nil, ErrInvalidFilename
+	}
+
+	// Try to initialize the shared module
+	api := pkcs11.New(module)
+	if err = api.Initialize(); err != nil {
+		return nil, ErrInitFailure
+	}
+	cleanup = func() {
+		api.Finalize()
+		api.Destroy()
+	}
+
+	// Get the list of slots.
+	slots, err := api.GetSlotList(true)
+	if err != nil {
+		cleanup()
+		return nil, ErrMissingSlot
+	}
+
+	// For now, just assume that we can use slot 0.  Ideally, we
+	// would either let the slot id be passed in, or use a
+	// specific name.
+	session, err := api.OpenSession(slots[0], pkcs11.CKF_SERIAL_SESSION|pkcs11.CKF_RW_SESSION)
+	if err != nil {
+		cleanup()
+		return nil, ErrSessionFailure
+	}
+
+	// Attempt to log into this slot using the provided PIN.
+	if err = api.Login(session, pkcs11.CKU_USER, pin); err != nil {
+		cleanup()
+		return nil, ErrLoginFailure
+	}
```

Comment:
I wonder if there are any multi-threading issues with a single connection for all requests?